UIforETW/Utility.cpp

```diff
@@ -20,7 +20,305 @@ limitations under the License.
 #include <direct.h>
 #include <ShlObj.h>
 
-std::vector<std::wstring> split(const std::wstring& s, char c)
+
+namespace {
+PCWSTR const WPAStartupFileName = L"\\Startup.wpaProfile";
+
+_Success_(return)
+bool AllocGlobalMemoryForString(const std::wstring& text, _Out_ _Post_valid_ HGLOBAL* const hmem)
+{
+	const size_t length = (text.size() + 1) * sizeof(text[0]);
+	const HANDLE mem = ::GlobalAlloc(GMEM_MOVEABLE, length);
+	if (!mem)
+	{
+		outputPrintf(L"Couldn't allocate memory to set clipboard text!\n");
+		outputLastError();
+		return false;
+	}
+	(*hmem) = mem;
+	return true;
+}
+
+void FreeGlobalMemory(_Frees_ptr_opt_ const HGLOBAL hmem)
+{
+		//If [GlobalFree] succeeds, the return value is NULL.
+		const HGLOBAL freeRes = ::GlobalFree(hmem);
+		if (freeRes != NULL)
+			std::terminate();//Logic bug!
+}
+
+_Success_(return)
+bool LockGlobalMemory(_In_ _Pre_valid_ const HGLOBAL hmem, _Outptr_ PVOID* const ptrMem)
+{
+	void* const ptr = ::GlobalLock(hmem);
+	if (ptr == NULL)
+	{
+		outputPrintf(L"Failed to lock memory for clipboard text!\n");
+		outputLastError();
+		return false;
+	}
+	(*ptrMem) = ptr;
+	return true;
+}
+
+void UnlockGlobalMemory(_In_ const HGLOBAL hmem)
+{
+	const BOOL unlockRes = ::GlobalUnlock(hmem);
+	if (!unlockRes)
+	{
+		const DWORD lastErr = ::GetLastError();
+		if (lastErr != NO_ERROR)
+			std::terminate( );//Logic bug!
+	}
+}
+
+_Success_(return)
+bool SizeGlobalMemory(_In_ _Pre_valid_ const HGLOBAL hmem, _Out_ size_t* const sizeBytes)
+{
+	const size_t bytes = ::GlobalSize(hmem);
+	if (bytes == 0)
+	{
+		outputPrintf(L"Failed to get size of globally-allocated memory!\n");
+		outputLastError();
+		return false;
+	}
+	(*sizeBytes) = bytes;
+	return true;
+}
+
+_Success_(return)
+bool ClipboardGetUnicode(_Out_ _Post_valid_ HANDLE* const hClipboard)
+{
+	const HANDLE hClip = ::GetClipboardData(CF_UNICODETEXT);
+	if (hClip == NULL)
+	{
+		outputPrintf(L"Failed to get clipboard data!\n");
+		outputLastError();
+		return false;
+	}
+	(*hClipboard) = hClip;
+	return true;
+}
+
+_Success_(return)
+bool ClipboardSetUnicode(_In_ const HGLOBAL hmem)
+{
+	const HANDLE dataHandle = ::SetClipboardData(CF_UNICODETEXT, hmem);
+	if (dataHandle == NULL)
+	{
+		outputPrintf(L"Failed to set clipboard data!\n");
+		outputLastError();
+		return false;
+	}
+	return true;
+}
+
+void closeClipboard()
+{
+	const BOOL closeResult = ::CloseClipboard( );
+	if (closeResult)
+		return;
+	debugPrintf(L"Failed to close the clipboard!\n");
+	debugLastError( );
+}
+
+bool openClipboard()
+{
+	const BOOL openClip = ::OpenClipboard(::GetDesktopWindow());
+	if (!openClip)
+	{
+		outputPrintf(L"Failed to open clipboard!\n");
+		debugLastError();
+		return false;
+	}
+	return true;
+}
+
+bool emptyClipboard()
+{
+	const BOOL emptyClip = ::EmptyClipboard();
+	if (!emptyClip)
+	{
+		outputPrintf(L"Failed to empty clipboard!\n");

```

Comment:
Logging is probably unnecessary (like above), but a failure to empty the clipboard will mean either:

1. `SetClipboardData` fails later, for mysterious reasons
2. The clipboard may - and this is a guess - end up with garbage data of some sort.

...should I also `std::terminate()` or `UIETWASSERT` this? (thus the function would become a `void` function).